library/init/algebra/group.lean

```diff
@@ -12,302 +12,312 @@ set_option default_priority 100
 
 set_option old_structure_cmd true
 
+/- Additive "sister" structures.
+   Example, add_semigroup mirrors semigroup.
+   These structures exist just to help automation.
+   In an alternative design, we could have the binary operation as an
+   extra argument for semigroup, monoid, group, etc. However, the lemmas
+   would be hard to index since they would not contain any constant.
+   For example, mul_assoc would be
+
+   lemma mul_assoc {α : Type u} {op : α → α → α} [semigroup α op] :
+                   ∀ a b c : α, op (op a b) c = op a (op b c) :=
+    semigroup.mul_assoc
+
+   The simplifier cannot effectively use this lemma since the pattern for
+   the left-hand-side would be
+
+        ?op (?op ?a ?b) ?c
+
+   Remark: we use a tactic for transporting theorems from the multiplicative fragment
+   to the additive one.
+-/
+
+/- Transport multiplicative to additive -/
+section transport
+open tactic
+
+meta def transport_with_dict (dict : name_map name) (src : name) (tgt : name) : command :=
+copy_decl_using dict src tgt
+>> copy_attribute `reducible src tt tgt
+>> copy_attribute `simp src tt tgt
+>> copy_attribute `instance src tt tgt
+
+@[user_attribute]
+meta def to_additive_attr : user_attribute (name_map name) name :=
+{ name      := `to_additive,
+  descr     := "Transport multiplicative to additive",
+  cache_cfg := ⟨λ ns, ns.mfoldl (λ dict n, do
+    val ← to_additive_attr.get_param n,
+    pure $ dict.insert n val) mk_name_map, []⟩,
+  parser    := lean.parser.ident,
+  after_set := some $ λ src _ _, do
+    env ← get_env,
+    dict ← to_additive_attr.get_cache,
```

Comment:
It would have been nice to have a `set_cache` function here, rather than having to rebuild the entire name map from scratch on every new theorem. Is this possible in the current architecture (that is, is possible for the user data of a user attr to be read/write rather than just derived from the list of declarations with the attribute)?

Reply:
That sounds like a weird API to me. The cache is supposed to transparently memoize the output of the caching function, explicitly setting it breaks that model. Or are you thinking of it as an "unsafe" "Trust me on this, I know what I'm doing" operation?
In any case, I would be surprised if it had a measurable performance impact for this use case.

Reply:
Hm, in this case it's still pure-functional, but it's hard to tell that it is because the data is coming from tactics. I think it should just be treated as an "unsafe" function, as you say. I agree that this probably won't be a considerable performance impact, it's a minor thing, but if I understand correctly the current implementation is O(n^2 log n) over n definitions, where it could be O(n log n).